### PR TITLE
Leaves - Mariya

### PR DESCRIPTION
**Pull request by M-Burr**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote custom methods for top ten in a category, spotlight, and listing all the works within a category by rating (highest to lowest).
Describe how you approached testing that model method. What edge cases did you come up with? | I wrote a test for what the spotlight feature will return if there are no works (and therefore no spotlight work to feature).
What are `session` and `flash`? What is the difference between them? | Session is a specific hash that we use to track a user's session & where the monitoring ends when the user has closed their browser. Flash is a different hash that allows you to send messages (that last 1 request-response cycle) from our controllers into our views.
What was one thing that you gained more clarity on through this assignment? | I have more clarity on sessions vs. views and how to use them.
What is the Heroku URL of your deployed application? |  https://mariyas-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | It would have been nice to have more flexibility to design our site.

### CODE REVIEW

File: app/views/works/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<main>
```
Comment (beccaelenzil) [inline app/views/works/edit.html.erb:1]: This would be a good place to use a partial view to DRY up your code for the new and edit views.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
+class UsersController < ApplicationController
+  
+  def index 
+    @users = User.all
+  end
+  
+  def show
+    user_id = params[:id]
+    @user = User.find_by(id: user_id)
```
Comment (beccaelenzil) [inline app/controllers/users_controller.rb:9]: This would be good functionality to put in a controllor filter `find_user` in the Application Controller (so that all controllers inherit this method).

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
+class WorksController < ApplicationController
+  
+  def index
+    @works = Work.all
+  end
+  
+  def show
+    work_id = params[:id]
+    @work = Work.find_by(id: work_id)
```
Comment (beccaelenzil) [inline app/controllers/works_controller.rb:9]: This would be good functionality to put in a controller filter -- a private method in the works controller called something like `find_work`.

File: test/models/user_test.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    before do 
+      # Arrange
+      @user = User.create(username: "hello_world")
+      @movie = Work.create(category: "movie", title: "Nightmare Before Chirstmas", creator: "Tim Burton", publication_year: 1993, description: "Jack Skellington tries to bring Christmas magic to Halloween Town. Chaos ensues.")
+      @movie2 = Work.create(category: "movie", title: "Hocus Pocus", creator: "Fox", publication_year: 1993, description: "Super spooky.")
+      @vote = Vote.create(user_id: @user.id, work_id: @movie.id)
+      @vote2 = Vote.create(user_id: @user.id, work_id: @movie2.id)
+    end
+    
+    it "can have many votes" do
```
Comment (beccaelenzil) [inline test/models/user_test.rb:38]: For an even more thorough tests, assert that @User.votes[0].id = @vote.id

**Comment by beccaelenzil** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | The flash message was "You already vote for this work." It should be something along the lines of "You must be logged in to vote."
Can edit this work successfully, and get a flash message | check
Can go to "View all media" page and see three lists of works, sorted by vote | check
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | check -- consider looping through the messages in your view to format your flash message extra politely :)
Can create a new work successfully. Note the URL for this work's show page | check
Can delete this work successfully |yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | yes -- consider a flash message to let the user know what happened.
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works |yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | see comment
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | see comment
Some controller, likely the `WorksController`, has a controller filter for finding a work | see comment
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | I see you didn't use fixtures. You have an opportunity to work on this in betsy.
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | check
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | check
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | check
Work model has tests with a section on all business logic methods in the model, including their edge cases | check

## Overall Feedback
Good work overall! Your implementation matches the demo site very closely, and the learning goals for this assignment were definitely met. I've left a few comments to review and consider. In coming projects, try to incorporate the use of fixtures, parital views, and controller filters.. Keep up the hard work!

File: app/views/homepages/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ <p class="spotlight__description"><%= feature.votes.count %> votes</p>
+</span>
+<% end %>
+</section>
+<hr class="root__hr">
+<div class="row">
+<section class="col-4">
+<h2 class="category">Top Movies</h2>
+<div>
+<% movies = @works.top_category("movie") %>
```
Comment (beccaelenzil) [inline app/views/homepages/index.html.erb:23]: Consider using a partial view to DRY up this repeated code.
